lib/charms/lego_base_k8s/v0/lego_client.py

```diff
@@ -191,34 +212,53 @@ def _on_certificate_creation_request(self, event: CertificateCreationRequestEven
         - Pulls certificates from workload
         - Sends certificates to requesting charm
         """
-        self._on_config_changed(event)
-        if not isinstance(self.unit.status, ActiveStatus):
-            event.defer()
+        if not self._is_unit_active():
+            logger.debug(
+                "Charm is not active, skipping certificate generation, \
+                will try again in during the next update status event."
+            )
             return
+        self._generate_signed_certificate(event.certificate_signing_request, event.relation_id)
+
+    def _generate_signed_certificate(self, csr: str, relation_id: int):
+        """Generate signed certificate from the ACME provider."""
         if not self.unit.is_leader():
+            logger.debug("Only the leader can handle certificate requests")
             return
         if not self._container.can_connect():
             logger.info("Waiting for container to be ready")
-            event.defer()
             return
-        csr_subject = self._get_subject_from_csr(event.certificate_signing_request)
+        csr_subject = self._get_subject_from_csr(csr)
         if len(csr_subject) > 64:
             logger.error("Subject is too long (> 64 characters): %s", csr_subject)
             return
         logger.info("Received Certificate Creation Request for domain %s", csr_subject)
-        self._push_csr_to_workload(event.certificate_signing_request)
+        self._push_csr_to_workload(csr=csr)
         if not self._execute_lego_cmd():
-            logger.error("Failed to execute lego command")
+            logger.error(
+                "Failed to execute lego command \
+                will try again in during the next update status event."
+            )
+            return
+        if not (signed_certificates := self._pull_certificates_from_workload(csr_subject)):
+            logger.error(
+                "Failed to pull certificates from workload \
+                will try again in during the next update status event."
+            )
             return
-        signed_certificates = self._pull_certificates_from_workload(csr_subject)
         self.tls_certificates.set_relation_certificate(
             certificate=signed_certificates[0],
-            certificate_signing_request=event.certificate_signing_request,
+            certificate_signing_request=csr,
             ca=signed_certificates[-1],
             chain=list(reversed(signed_certificates)),
-            relation_id=event.relation_id,
+            relation_id=relation_id,
         )
 
+    def _is_unit_active(self) -> bool:
+        """Check if the unit is active."""
+        self._on_config_changed(None)
```

Comment:
I agree, but since behavior is not introduced in this PR I prefer not to mix the design change with fixing this "bug". I created a ticket and started working on a separate PR.